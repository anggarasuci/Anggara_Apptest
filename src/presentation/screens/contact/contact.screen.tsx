import React, {useCallback} from 'react';
import {ContactViewModel} from './contact.viewmodel';
import {FlatList, Text, TouchableOpacity, View} from 'react-native';
import {ContactModel} from '../../../domain/models/contact.model';
import CardView from '../../components/card/card';
import CircleImageView from '../../components/circle-image/circle.image';
import {contactStyle as styles} from './contact.style';
import VerticalTextView from '../../components/vertical-text/vertical.text';
import ConfirmationDialog from '../../components/dialog/confirmation.dialog';

const ContactScreen = () => {
  const {
    items,
    isDeleteDialogVisible,
    setIsDeleteDialogVisible,
    onDeleteClicked,
    onConfirmDelete,
    onNavigateToCreatePage,
  } = ContactViewModel();

  const getContent = useCallback((item: ContactModel) => {
    return (
      <CardView onClick={() => onNavigateToCreatePage(item.id)}>
        <View style={styles.removeContainer}>
          <TouchableOpacity
            style={styles.removeButton}
            onPress={() => onDeleteClicked(item.id)}>
            <Text style={styles.removeText}>X</Text>
          </TouchableOpacity>
        </View>
        <View style={styles.card}>
          <CircleImageView url={item.photo} />
          <View style={styles.hSpacer} />
          <View>
            <VerticalTextView label="Nama Lengkap" value={item.fullName} />
            <View style={styles.spacer} />
            <VerticalTextView label="Umur" value={item.age.toString()} />
          </View>
        </View>
      </CardView>
    );
  }, []);

  return (
    <>
      <View>
        {items?.length > 0 && (
          <FlatList
            data={items}
            keyExtractor={(item: ContactModel) => item.id.toString()}
            renderItem={({item}) => getContent(item)}
          />
        )}
      </View>
      <View style={styles.fabContainer}>
        <TouchableOpacity
          style={styles.fab}
          onPress={() => onNavigateToCreatePage()}>
          <Text style={styles.fabText}>+</Text>
        </TouchableOpacity>
      </View>

      <ConfirmationDialog
        label={'Are you sure want to delete this item?'}
        isVisible={isDeleteDialogVisible}
        onCancel={() => setIsDeleteDialogVisible(false)}
        onConfirm={onConfirmDelete}
      />
    </>
  );
};

export default ContactScreen;
