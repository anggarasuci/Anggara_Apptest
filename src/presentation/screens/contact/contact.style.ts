import {StyleSheet} from 'react-native';

export const contactStyle = StyleSheet.create({
  card: {
    flexDirection: 'row',
  },
  spacer: {
    height: 8,
  },
  hSpacer: {
    width: 8,
  },
});
