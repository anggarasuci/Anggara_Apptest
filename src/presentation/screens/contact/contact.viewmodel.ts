import {useCallback, useEffect, useState} from 'react';
import ContactRepositoryImpl from '../../../data/repositories-impl/contact.repository.impl';
import ContactUseCase from '../../../domain/use-case/contact.usecase';
import {useDispatch, useSelector} from 'react-redux';
import ContactAction from '../../../infrastructure/redux/actions/contact.action';
import {RootState} from '../../../infrastructure/redux/store';
import {ContactState} from '../../../infrastructure/redux/reducers/contact.reducer';
import {useFocusEffect, useNavigation} from '@react-navigation/native';
import Toast from 'react-native-root-toast';

const ContactViewModel = () => {
  const contactRepository: ContactRepositoryImpl = new ContactRepositoryImpl();
  const contactUseCase: ContactUseCase = new ContactUseCase(contactRepository);
  const {setContacts, refreshContacts, deleteContact} = ContactAction();
  const dispatch = useDispatch();
  const state: ContactState = useSelector((state: RootState) => state.contact);
  const navigation = useNavigation();

  const [isDeleteDialogVisible, setIsDeleteDialogVisible] = useState(false);
  const [selectedId, setSelectedId] = useState('');

  useEffect(() => {
    getItems();
  }, []);

  useFocusEffect(() => {
    if (state?.isRefreshState) {
      getItems();
      dispatch(refreshContacts(false));
    }
  });

  const getItems = async () => {
    const result = await contactUseCase.getItems();
    dispatch(setContacts(result));
  };

  const onNavigateToCreatePage = useCallback((id?: string) => {
    navigation.navigate('Create Contact', {id: id});
  }, []);

  const onDeleteClicked = useCallback((id: string) => {
    setSelectedId(id);
    setIsDeleteDialogVisible(true);
  }, []);

  const onConfirmDelete = useCallback(async () => {
    const result = await contactUseCase.delete(selectedId);
    Toast.show(result.error ?? 'Success Deleted', {
      duration: Toast.durations.SHORT,
    });

    if (!result?.error) {
      dispatch(deleteContact(selectedId));
    }
    setIsDeleteDialogVisible(false);
  }, [selectedId, dispatch]);

  return {
    items: state?.items ?? [],
    isDeleteDialogVisible,
    onNavigateToCreatePage,
    onDeleteClicked,
    onConfirmDelete,
    setIsDeleteDialogVisible,
  };
};

export {ContactViewModel};
