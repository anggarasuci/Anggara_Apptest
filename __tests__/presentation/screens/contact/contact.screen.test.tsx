import React from 'react';
import {render} from '@testing-library/react-native';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import {NavigationContainer} from '@react-navigation/native';
import ContactScreen from '../../../../src/presentation/screens/contact/contact.screen'; // Import the component you want to test
import {initialState} from '../../../../src/infrastructure/redux/reducers/contact.reducer'; // Import your initial Redux state

const mockStore = configureStore([]);

test('renders ContactScreen correctly', () => {
  const store = mockStore(initialState);

  const {toJSON} = render(
    <Provider store={store}>
      <NavigationContainer>
        <ContactScreen />
      </NavigationContainer>
    </Provider>,
  );

  expect(toJSON()).toMatchSnapshot();
});
