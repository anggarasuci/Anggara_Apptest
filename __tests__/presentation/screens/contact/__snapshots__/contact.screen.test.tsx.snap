// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ContactScreen correctly 1`] = `
[
  <View />,
  <View
    style={
      {
        "bottom": 16,
        "position": "absolute",
        "right": 16,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#007bff",
          "borderRadius": 28,
          "elevation": 8,
          "height": 56,
          "justifyContent": "center",
          "opacity": 1,
          "width": 56,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 24,
          }
        }
      >
        +
      </Text>
    </View>
  </View>,
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />,
]
`;
